### PR TITLE
Harvest overcloud.json after deployment

### PR DESCRIPTION
**Pull request by dprince**
This patch harvests overcloud.json (generated by TripleO Ansible)
after a successful deployment finishes. The results are merged
into other "exported" data from the Heat stack and then written
to the 'tripleo-ctlplane-exports' ConfigMap.

Then intent is to optionally use this ConfigMap alongside
a new ConfigGenerator to implement a multi-stack workflows
like DCN.

### CODE REVIEW

File: controllers/openstackconfiggenerator_controller.go [hunk 1]
```diff
@@ -683,10 +686,32 @@ func (r *OpenStackConfigGeneratorReconciler) verifyNodeResourceStatus(instance *
...
 		}
 	}
+	hasControllersForRoles := r.isControllerGenerator(vmset, bmset)
 
 	return msg, true, nil
 }
 
+//
+// Is this ConfigGenerator creating resources for a Controller
+//
```
Comment (dprince) [inline controllers/openstackconfiggenerator_controller.go:696]: This part is still in progress and perhaps I should break it out into a separate patch. The intent is to use this function to only execute the overcloud.json harvesting if the stack *is* the ControlPlane (i.e. has Controller roles in it)

File: containers/agent/deploy.go [hunk 2]
```diff
@@ -41,13 +49,29 @@ var (
...
+		Resource: "openstackconfigversions",
+	}
+
+	filteredAllNodesConfig = []string{
+		"oslo_messaging_notify_short_bootstrap_node_name",
+		"oslo_messaging_notify_node_names",
+		"oslo_messaging_rpc_node_names",
+		"memcached_node_ips",
+		"ovn_dbs_vip",
+		"redis_vip",
```
Comment (stuggi) [inline containers/agent/deploy.go:65]: wondering if this should be a parameter of the osdeploy (defaulting to this list) and then passed into the agent, or the agent get it from checking the osdeploy CR? with that this list could be easy customized in case filter requirements change.

Also filter is only done for a cell export [1] from the central controller deployment stack. In case of splitting the cell controllers and cell computes in different stacks, you use the same export like DCN, where there is no filtering [2], also not for DCN export [3]. So like we discussed in one of our previous meetings, we probably should just do both exports, one which filters and the one without?

[1] https://github.com/openstack/python-tripleoclient/blob/stable/train/tripleoclient/v1/overcloud_cell.py#L89
[2] https://github.com/openstack/python-tripleoclient/blob/stable/train/tripleoclient/v1/overcloud_cell.py#L92
[3] https://github.com/openstack/python-tripleoclient/blob/stable/train/tripleoclient/v1/overcloud_export.py#L93

File: containers/agent/deploy.go [hunk 3]
```diff
@@ -93,6 +190,97 @@ func ExecPodCommand(kclient kubernetes.Clientset, pod corev1.Pod, containerName
...
+		context.TODO(),
+		deployOpts.pod,
+		metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// obtain the allNodesData from the overcloud.json from the deployment pod
+	// this is written by TripleO's Ansible after successful deployment
+	allNodesDataString, err := CopyFromPod(kclient, *pod, "openstackclient", "/home/cloud-admin/work/"+deployOpts.configVersion+"/playbooks/tripleo-ansible/group_vars/overcloud.json")
```
Comment (stuggi) [inline containers/agent/deploy.go:239]: I do not have an env right now, does this also have these details?
```
        "EndpointMap": {
            "parameter": "EndpointMapOverride",
        },
        "HostsEntry": {
            "parameter": "ExtraHostFileEntries",
        },
        "GlobalConfig": {
            "parameter": "GlobalConfigExtraMapData",
        },
```

File: api/v1beta1/openstackconfigversion_types.go [hunk 4]
```diff
@@ -36,7 +36,7 @@ type OpenStackConfigVersionStatus struct {
 //+kubebuilder:subresource:status
 // +kubebuilder:resource:shortName=osconfigversion;osconfigversions
 // +operator-sdk:csv:customresourcedefinitions:displayName="OpenStack Config Version"
-// +kubebuilder:printcolumn:name="Generator",type="string",JSONPath=".spec.ConfigGeneratorName",description="Config Generator Name"
+// +kubebuilder:printcolumn:name="Generator",type="string",JSONPath=".spec.configGeneratorName",description="Config Generator Name"
```
Comment (dprince) [inline api/v1beta1/openstackconfigversion_types.go:39]: Initially I had these things in the same patch. This fixes the name which needs to be camelcase. Not sure how I didn't copy that over to the 2nd review which already landed.

File: containers/agent/deploy.go (hunk 2 above)
Comment (dprince) [reply containers/agent/deploy.go:65]: @stuggi just pushed a commit which splits out both filtered and non-filtered version of the file in the ConfigMap. Still need to add documentation on this to the README.
On the idea of making it a parameter while we could do that I suppose I'd rather not. Tripleclient has these as a constant so I'm not sure they change that often.

File: containers/agent/deploy.go (hunk 3 above)
Comment (dprince) [reply containers/agent/deploy.go:242]: Well, it did last time I tested. These take a while to test as you have to wait for a full TripleO deploy to finish. Trying it again this morning.

Comment (dprince) [reply containers/agent/deploy.go:242]: yes, these are all in the resulting environment file

File: README.md [hunk 5]
```diff
@@ -1545,6 +1545,25 @@ openstackbackuprestore   restore    openstackbackupsave-1641928378   Restored
...
 
+Exported stack data config map
+--------------------------------------------------------
+
+The OSP Director Operator automatically creates a ConfigMap after each OSDeploy resource finishes executing. This ConfigMap is named after the OSDeploy resource name and prefixed with tripleo-exports-. For example tripleo-exports-default would be the name of the ConfigMap for the 'default' OSDeploy resource. Each ConfigMap contains 2 YAML files:
+
+| Filename    | Description | TripleO Command Equivalent |
+| ----------- | ----------- | -------------------------- |
+| ctlplane-export.yaml | Used with multiple stacks for DCN | overcloud export |
+| ctlplane-export-filtered.yaml | Used for multiple stacks with Cells | overcloud cell export |
```
Comment (stuggi) [inline README.md:1556]: maybe we want to be more specific that this is used for cell stacks where the cell controllers get installed. if you split out again cell computes from a cell controller stack you'd use the ctlplane-export.yaml from the cell controller stack.

File: containers/agent/deploy.go [hunk 6]
```diff
@@ -93,6 +191,111 @@ func ExecPodCommand(kclient kubernetes.Clientset, pod corev1.Pod, containerName
...
```
Comment (stuggi) [inline containers/agent/deploy.go:242]: not required in this patch, but we might also want to get the clouds.yaml and /etc/hosts from the deployment. in case there is an issue with the storage of the openstackclient, thats some useful information to restore it.

**Review by stuggi — COMMENTED** [review]
just two minor comments, other then that it looks good to me.

File: containers/agent/deploy.go (hunk 6 above)
Comment (dprince) [reply containers/agent/deploy.go:242]: @stuggi yeah, maybe we file this as Jira tickets and can follow up on it as a "core" issue.

File: README.md (hunk 5 above)
Comment (dprince) [reply README.md]: Sure, I added a note to indicate the cell stacks as "Controllers"

**Review by stuggi — APPROVED** [review]
/lgtm
